File: dp/cpp/bridge.cpp
#include <iostream>
#include <memory>
#include <string>
#include <vector>
#include <sstream>

// Implementation interface (Bridge)
class DrawingAPI {
public:
    virtual ~DrawingAPI() = default;
    virtual void drawLine(double x1, double y1, double x2, double y2) = 0;
    virtual void drawCircle(double x, double y, double radius) = 0;
    virtual void drawRectangle(double x, double y, double width, double height) = 0;
    virtual void setColor(const std::string& color) = 0;
    virtual std::string getRendererName() const = 0;
};

// Concrete Implementation 1: OpenGL
class OpenGLRenderer : public DrawingAPI {
private:
    std::string currentColor = "white";

public:
    void drawLine(double x1, double y1, double x2, double y2) override {
        std::cout << "[OpenGL] Drawing line from (" << x1 << "," << y1
                  << ") to (" << x2 << "," << y2 << ") in " << currentColor << std::endl;
    }

    void drawCircle(double x, double y, double radius) override {
        std::cout << "[OpenGL] Drawing circle at (" << x << "," << y
                  << ") with radius " << radius << " in " << currentColor << std::endl;
    }

    void drawRectangle(double x, double y, double width, double height) override {
        std::cout << "[OpenGL] Drawing rectangle at (" << x << "," << y
                  << ") size " << width << "x" << height << " in " << currentColor << std::endl;
    }

    void setColor(const std::string& color) override {
        currentColor = color;
        std::cout << "[OpenGL] Color set to " << color << std::endl;
    }

    std::string getRendererName() const override {
        return "OpenGL";
    }
};

// Concrete Implementation 2: DirectX
class DirectXRenderer : public DrawingAPI {
private:
    std::string currentColor = "white";

public:
    void drawLine(double x1, double y1, double x2, double y2) override {
        std::cout << "[DirectX] Rendering line: (" << x1 << "," << y1
                  << ")->(" << x2 << "," << y2 << ") color=" << currentColor << std::endl;
    }

    void drawCircle(double x, double y, double radius) override {
        std::cout << "[DirectX] Rendering circle: center=(" << x << "," << y
                  << ") r=" << radius << " color=" << currentColor << std::endl;
    }

    void drawRectangle(double x, double y, double width, double height) override {
        std::cout << "[DirectX] Rendering rectangle: pos=(" << x << "," << y
                  << ") size=" << width << "x" << height << " color=" << currentColor << std::endl;
    }

    void setColor(const std::string& color) override {
        currentColor = color;
        std::cout << "[DirectX] Color changed to " << color << std::endl;
    }

    std::string getRendererName() const override {
        return "DirectX";
    }
};

// Concrete Implementation 3: SVG
class SVGRenderer : public DrawingAPI {
private:
    std::string currentColor = "black";
    std::ostringstream svgContent;

public:
    void drawLine(double x1, double y1, double x2, double y2) override {
        svgContent << "<line x1=\"" << x1 << "\" y1=\"" << y1
                   << "\" x2=\"" << x2 << "\" y2=\"" << y2
                   << "\" stroke=\"" << currentColor << "\"/>\n";
        std::cout << "[SVG] Added line element" << std::endl;
    }

    void drawCircle(double x, double y, double radius) override {
        svgContent << "<circle cx=\"" << x << "\" cy=\"" << y
                   << "\" r=\"" << radius << "\" fill=\"" << currentColor << "\"/>\n";
        std::cout << "[SVG] Added circle element" << std::endl;
    }

    void drawRectangle(double x, double y, double width, double height) override {
        svgContent << "<rect x=\"" << x << "\" y=\"" << y
                   << "\" width=\"" << width << "\" height=\"" << height
                   << "\" fill=\"" << currentColor << "\"/>\n";
        std::cout << "[SVG] Added rectangle element" << std::endl;
    }

    void setColor(const std::string& color) override {
        currentColor = color;
        std::cout << "[SVG] Color set to " << color << std::endl;
    }

    std::string getRendererName() const override {
        return "SVG";
    }

    std::string getSVGContent() const {
        return "<svg>\n" + svgContent.str() + "</svg>";
    }
};

// Abstraction: Shape
class Shape {
protected:
    std::unique_ptr<DrawingAPI> drawingAPI;
    double x, y;

public:
    Shape(std::unique_ptr<DrawingAPI> api, double x = 0, double y = 0)
        : drawingAPI(std::move(api)), x(x), y(y) {}

    virtual ~Shape() = default;
    virtual void draw() = 0;
    virtual void move(double newX, double newY) {
        x = newX;
        y = newY;
    }

    void setColor(const std::string& color) {
        drawingAPI->setColor(color);
    }

    std::string getRendererName() const {
        return drawingAPI->getRendererName();
    }
};

// Refined Abstraction: Circle
class Circle : public Shape {
private:
    double radius;

public:
    Circle(std::unique_ptr<DrawingAPI> api, double x, double y, double radius)
        : Shape(std::move(api), x, y), radius(radius) {}

    void draw() override {
        drawingAPI->drawCircle(x, y, radius);
    }

    void setRadius(double newRadius) {
        radius = newRadius;
    }

    double getRadius() const {
        return radius;
    }
};

// Refined Abstraction: Rectangle
class Rectangle : public Shape {
private:
    double width, height;

public:
    Rectangle(std::unique_ptr<DrawingAPI> api, double x, double y, double width, double height)
        : Shape(std::move(api), x, y), width(width), height(height) {}

    void draw() override {
        drawingAPI->drawRectangle(x, y, width, height);
    }

    void resize(double newWidth, double newHeight) {
        width = newWidth;
        height = newHeight;
    }
};

// Refined Abstraction: Line
class Line : public Shape {
private:
    double x2, y2;

public:
    Line(std::unique_ptr<DrawingAPI> api, double x1, double y1, double x2, double y2)
        : Shape(std::move(api), x1, y1), x2(x2), y2(y2) {}

    void draw() override {
        drawingAPI->drawLine(x, y, x2, y2);
    }

    void setEndPoint(double newX2, double newY2) {
        x2 = newX2;
        y2 = newY2;
    }
};

// Complex shape composed of basic shapes
class House : public Shape {
private:
    std::unique_ptr<Rectangle> base;
    std::unique_ptr<Rectangle> roof;
    std::unique_ptr<Rectangle> door;
    std::unique_ptr<Circle> window;

public:
    House(std::unique_ptr<DrawingAPI> api, double x, double y)
        : Shape(std::move(api), x, y) {
        // Create a copy of the drawing API for each component
        // Note: In a real implementation, you might want to share the API or use a different approach
    }

    void draw() override {
        std::cout << "Drawing house at (" << x << "," << y << ") using "
                  << drawingAPI->getRendererName() << std::endl;

        // Draw base
        drawingAPI->setColor("brown");
        drawingAPI->drawRectangle(x, y, 100, 80);

        // Draw roof
        drawingAPI->setColor("red");
        drawingAPI->drawRectangle(x - 10, y + 80, 120, 40);

        // Draw door
        drawingAPI->setColor("darkbrown");
        drawingAPI->drawRectangle(x + 40, y, 20, 50);

        // Draw window
        drawingAPI->setColor("lightblue");
        drawingAPI->drawCircle(x + 20, y + 60, 8);
        drawingAPI->drawCircle(x + 80, y + 60, 8);
    }
};

// Example with different rendering backends for notification system
class NotificationSender {
public:
    virtual ~NotificationSender() = default;
    virtual void sendNotification(const std::string& title, const std::string& message) = 0;
    virtual std::string getChannelName() const = 0;
};

class EmailSender : public NotificationSender {
public:
    void sendNotification(const std::string& title, const std::string& message) override {
        std::cout << "[EMAIL] To: user@example.com" << std::endl;
        std::cout << "[EMAIL] Subject: " << title << std::endl;
        std::cout << "[EMAIL] Body: " << message << std::endl;
    }

    std::string getChannelName() const override {
        return "Email";
    }
};

class SMSSender : public NotificationSender {
public:
    void sendNotification(const std::string& title, const std::string& message) override {
        std::cout << "[SMS] " << title << ": " << message << " (160 char limit)" << std::endl;
    }

    std::string getChannelName() const override {
        return "SMS";
    }
};

class PushNotificationSender : public NotificationSender {
public:
    void sendNotification(const std::string& title, const std::string& message) override {
        std::cout << "[PUSH] 📱 " << title << std::endl;
        std::cout << "[PUSH] " << message << std::endl;
    }

    std::string getChannelName() const override {
        return "Push Notification";
    }
};

// Abstraction for notifications
class Notification {
protected:
    std::unique_ptr<NotificationSender> sender;
    std::string title;
    std::string message;

public:
    Notification(std::unique_ptr<NotificationSender> sender,
                const std::string& title, const std::string& message)
        : sender(std::move(sender)), title(title), message(message) {}

    virtual ~Notification() = default;
    virtual void send() = 0;

    std::string getChannelName() const {
        return sender->getChannelName();
    }
};

// Simple notification
class SimpleNotification : public Notification {
public:
    SimpleNotification(std::unique_ptr<NotificationSender> sender,
                      const std::string& title, const std::string& message)
        : Notification(std::move(sender), title, message) {}

    void send() override {
        sender->sendNotification(title, message);
    }
};

// Urgent notification with different formatting
class UrgentNotification : public Notification {
public:
    UrgentNotification(std::unique_ptr<NotificationSender> sender,
                      const std::string& title, const std::string& message)
        : Notification(std::move(sender), title, message) {}

    void send() override {
        std::string urgentTitle = "🚨 URGENT: " + title;
        std::string urgentMessage = "⚠️ " + message + " ⚠️";
        sender->sendNotification(urgentTitle, urgentMessage);
    }
};

int main() {
    std::cout << "=== Bridge Pattern Demo ===" << std::endl;

    // Example 1: Graphics rendering with different APIs
    std::cout << "\n1. Graphics Rendering Bridge:" << std::endl;
    std::cout << std::string(40, '-') << std::endl;

    std::vector<std::unique_ptr<Shape>> shapes;

    // Create shapes with different rendering backends
    shapes.push_back(std::make_unique<Circle>(
        std::make_unique<OpenGLRenderer>(), 10, 10, 5));
    shapes.push_back(std::make_unique<Rectangle>(
        std::make_unique<DirectXRenderer>(), 20, 20, 15, 10));
    shapes.push_back(std::make_unique<Line>(
        std::make_unique<SVGRenderer>(), 0, 0, 50, 50));

    for (auto& shape : shapes) {
        std::cout << "\nUsing " << shape->getRendererName() << ":" << std::endl;
        shape->setColor("blue");
        shape->draw();
    }

    // Complex shape example
    std::cout << "\n2. Complex Shape (House):" << std::endl;
    std::cout << std::string(40, '-') << std::endl;

    auto house1 = std::make_unique<House>(std::make_unique<OpenGLRenderer>(), 100, 100);
    auto house2 = std::make_unique<House>(std::make_unique<DirectXRenderer>(), 200, 100);

    house1->draw();
    std::cout << std::endl;
    house2->draw();

    // SVG content example
    std::cout << "\n3. SVG Content Generation:" << std::endl;
    std::cout << std::string(40, '-') << std::endl;

    auto svgRenderer = std::make_unique<SVGRenderer>();
    auto svgRendererPtr = svgRenderer.get(); // Keep a pointer for later use

    auto svgCircle = std::make_unique<Circle>(std::move(svgRenderer), 50, 50, 25);
    svgCircle->setColor("red");
    svgCircle->draw();

    std::cout << "\nGenerated SVG:" << std::endl;
    std::cout << svgRendererPtr->getSVGContent() << std::endl;

    // Example 2: Notification system bridge
    std::cout << "\n4. Notification System Bridge:" << std::endl;
    std::cout << std::string(40, '-') << std::endl;

    std::vector<std::unique_ptr<Notification>> notifications;

    // Create different types of notifications with different senders
    notifications.push_back(std::make_unique<SimpleNotification>(
        std::make_unique<EmailSender>(),
        "Welcome",
        "Thank you for signing up!"));

    notifications.push_back(std::make_unique<UrgentNotification>(
        std::make_unique<SMSSender>(),
        "Security Alert",
        "Suspicious login detected"));

    notifications.push_back(std::make_unique<SimpleNotification>(
        std::make_unique<PushNotificationSender>(),
        "New Message",
        "You have 3 unread messages"));

    notifications.push_back(std::make_unique<UrgentNotification>(
        std::make_unique<EmailSender>(),
        "System Maintenance",
        "Service will be down for 2 hours"));

    for (auto& notification : notifications) {
        std::cout << "\nSending via " << notification->getChannelName() << ":" << std::endl;
        notification->send();
    }

    // Runtime renderer switching example
    std::cout << "\n5. Runtime Renderer Switching:" << std::endl;
    std::cout << std::string(40, '-') << std::endl;

    auto circle = std::make_unique<Circle>(std::make_unique<OpenGLRenderer>(), 25, 25, 10);

    std::cout << "Original renderer:" << std::endl;
    circle->setColor("green");
    circle->draw();

    // Note: In a real implementation, you might want to provide a method to change the renderer
    // For demonstration, we'll create a new circle with a different renderer
    std::cout << "\nSwitching to DirectX renderer:" << std::endl;
    auto circle2 = std::make_unique<Circle>(std::make_unique<DirectXRenderer>(), 25, 25, 10);
    circle2->setColor("green");
    circle2->draw();

    return 0;
}

File: dp/rust/src/bridge.rs
/// Bridge Pattern Implementation in Rust
///
/// The Bridge pattern separates an abstraction from its implementation,
/// allowing both to vary independently. In Rust, this is implemented
/// using traits for the implementation interface and structs for abstractions.

use std::collections::HashMap;

/// Implementation interface for drawing operations
pub trait DrawingAPI {
    fn draw_circle(&self, x: f64, y: f64, radius: f64) -> String;
    fn draw_rectangle(&self, x: f64, y: f64, width: f64, height: f64) -> String;
    fn draw_line(&self, x1: f64, y1: f64, x2: f64, y2: f64) -> String;
    fn get_api_name(&self) -> &str;
}

/// Concrete implementation for Canvas API
pub struct CanvasAPI {
    name: String,
    canvas_size: (u32, u32),
}

impl CanvasAPI {
    pub fn new(width: u32, height: u32) -> Self {
        Self {
            name: "Canvas API".to_string(),
            canvas_size: (width, height),
        }
    }
}

impl DrawingAPI for CanvasAPI {
    fn draw_circle(&self, x: f64, y: f64, radius: f64) -> String {
        format!(
            "Canvas: Drawing circle at ({:.1}, {:.1}) with radius {:.1}",
            x, y, radius
        )
    }

    fn draw_rectangle(&self, x: f64, y: f64, width: f64, height: f64) -> String {
        format!(
            "Canvas: Drawing rectangle at ({:.1}, {:.1}) with size {:.1}x{:.1}",
            x, y, width, height
        )
    }

    fn draw_line(&self, x1: f64, y1: f64, x2: f64, y2: f64) -> String {
        format!(
            "Canvas: Drawing line from ({:.1}, {:.1}) to ({:.1}, {:.1})",
            x1, y1, x2, y2
        )
    }

    fn get_api_name(&self) -> &str {
        &self.name
    }
}

/// Concrete implementation for SVG API
pub struct SvgAPI {
    name: String,
    precision: u32,
}

impl SvgAPI {
    pub fn new(precision: u32) -> Self {
        Self {
            name: "SVG API".to_string(),
            precision,
        }
    }
}

impl DrawingAPI for SvgAPI {
    fn draw_circle(&self, x: f64, y: f64, radius: f64) -> String {
        format!(
            "<circle cx=\"{:.prec$}\" cy=\"{:.prec$}\" r=\"{:.prec$}\" />",
            x,
            y,
            radius,
            prec = self.precision as usize
        )
    }

    fn draw_rectangle(&self, x: f64, y: f64, width: f64, height: f64) -> String {
        format!(
            "<rect x=\"{:.prec$}\" y=\"{:.prec$}\" width=\"{:.prec$}\" height=\"{:.prec$}\" />",
            x,
            y,
            width,
            height,
            prec = self.precision as usize
        )
    }

    fn draw_line(&self, x1: f64, y1: f64, x2: f64, y2: f64) -> String {
        format!(
            "<line x1=\"{:.prec$}\" y1=\"{:.prec$}\" x2=\"{:.prec$}\" y2=\"{:.prec$}\" />",
            x1,
            y1,
            x2,
            y2,
            prec = self.precision as usize
        )
    }

    fn get_api_name(&self) -> &str {
        &self.name
    }
}

/// Concrete implementation for OpenGL API
pub struct OpenGLAPI {
    name: String,
    version: String,
}

impl OpenGLAPI {
    pub fn new(version: &str) -> Self {
        Self {
            name: "OpenGL API".to_string(),
            version: version.to_string(),
        }
    }
}

impl DrawingAPI for OpenGLAPI {
    fn draw_circle(&self, x: f64, y: f64, radius: f64) -> String {
        format!(
            "OpenGL {}: glBegin(GL_TRIANGLE_FAN); glVertex2f({:.2}, {:.2}); /* circle with radius {:.2} */",
            self.version, x, y, radius
        )
    }

    fn draw_rectangle(&self, x: f64, y: f64, width: f64, height: f64) -> String {
        format!(
            "OpenGL {}: glBegin(GL_QUADS); glVertex2f({:.2}, {:.2}); /* {}x{} rectangle */",
            self.version, x, y, width, height
        )
    }

    fn draw_line(&self, x1: f64, y1: f64, x2: f64, y2: f64) -> String {
        format!(
            "OpenGL {}: glBegin(GL_LINES); glVertex2f({:.2}, {:.2}); glVertex2f({:.2}, {:.2});",
            self.version, x1, y1, x2, y2
        )
    }

    fn get_api_name(&self) -> &str {
        &self.name
    }
}

/// Abstraction - Base shape class
pub trait Shape {
    fn draw(&self) -> Vec<String>;
    fn resize(&mut self, factor: f64);
    fn move_to(&mut self, x: f64, y: f64);
    fn get_area(&self) -> f64;
    fn get_info(&self) -> String;
}

/// Refined Abstraction - Circle
pub struct Circle {
    x: f64,
    y: f64,
    radius: f64,
    drawing_api: Box<dyn DrawingAPI>,
}

impl Circle {
    pub fn new(x: f64, y: f64, radius: f64, drawing_api: Box<dyn DrawingAPI>) -> Self {
        Self {
            x,
            y,
            radius,
            drawing_api,
        }
    }
}

impl Shape for Circle {
    fn draw(&self) -> Vec<String> {
        vec![self.drawing_api.draw_circle(self.x, self.y, self.radius)]
    }

    fn resize(&mut self, factor: f64) {
        self.radius *= factor;
    }

    fn move_to(&mut self, x: f64, y: f64) {
        self.x = x;
        self.y = y;
    }

    fn get_area(&self) -> f64 {
        std::f64::consts::PI * self.radius * self.radius
    }

    fn get_info(&self) -> String {
        format!(
            "Circle at ({:.1}, {:.1}) with radius {:.1} using {}",
            self.x,
            self.y,
            self.radius,
            self.drawing_api.get_api_name()
        )
    }
}

/// Refined Abstraction - Rectangle
pub struct Rectangle {
    x: f64,
    y: f64,
    width: f64,
    height: f64,
    drawing_api: Box<dyn DrawingAPI>,
}

impl Rectangle {
    pub fn new(x: f64, y: f64, width: f64, height: f64, drawing_api: Box<dyn DrawingAPI>) -> Self {
        Self {
            x,
            y,
            width,
            height,
            drawing_api,
        }
    }
}

impl Shape for Rectangle {
    fn draw(&self) -> Vec<String> {
        vec![self
            .drawing_api
            .draw_rectangle(self.x, self.y, self.width, self.height)]
    }

    fn resize(&mut self, factor: f64) {
        self.width *= factor;
        self.height *= factor;
    }

    fn move_to(&mut self, x: f64, y: f64) {
        self.x = x;
        self.y = y;
    }

    fn get_area(&self) -> f64 {
        self.width * self.height
    }

    fn get_info(&self) -> String {
        format!(
            "Rectangle at ({:.1}, {:.1}) with size {:.1}x{:.1} using {}",
            self.x,
            self.y,
            self.width,
            self.height,
            self.drawing_api.get_api_name()
        )
    }
}

/// Complex shape that uses multiple drawing operations
pub struct ComplexShape {
    components: Vec<String>,
    drawing_api: Box<dyn DrawingAPI>,
    center_x: f64,
    center_y: f64,
    scale: f64,
}

impl ComplexShape {
    pub fn new(center_x: f64, center_y: f64, drawing_api: Box<dyn DrawingAPI>) -> Self {
        Self {
            components: Vec::new(),
            drawing_api,
            center_x,
            center_y,
            scale: 1.0,
        }
    }

    /// Draw a house-like shape
    pub fn draw_house(&mut self) -> Vec<String> {
        let mut commands = Vec::new();

        // Base of house (rectangle)
        commands.push(self.drawing_api.draw_rectangle(
            self.center_x - 50.0 * self.scale,
            self.center_y - 30.0 * self.scale,
            100.0 * self.scale,
            60.0 * self.scale,
        ));

        // Roof (triangle made with lines)
        commands.push(self.drawing_api.draw_line(
            self.center_x - 60.0 * self.scale,
            self.center_y - 30.0 * self.scale,
            self.center_x,
            self.center_y - 80.0 * self.scale,
        ));

        commands.push(self.drawing_api.draw_line(
            self.center_x,
            self.center_y - 80.0 * self.scale,
            self.center_x + 60.0 * self.scale,
            self.center_y - 30.0 * self.scale,
        ));

        // Door (rectangle)
        commands.push(self.drawing_api.draw_rectangle(
            self.center_x - 10.0 * self.scale,
            self.center_y + 5.0 * self.scale,
            20.0 * self.scale,
            25.0 * self.scale,
        ));

        // Window (circle)
        commands.push(self.drawing_api.draw_circle(
            self.center_x + 25.0 * self.scale,
            self.center_y - 10.0 * self.scale,
            8.0 * self.scale,
        ));

        self.components = commands.clone();
        commands
    }
}

impl Shape for ComplexShape {
    fn draw(&self) -> Vec<String> {
        self.components.clone()
    }

    fn resize(&mut self, factor: f64) {
        self.scale *= factor;
    }

    fn move_to(&mut self, x: f64, y: f64) {
        self.center_x = x;
        self.center_y = y;
    }

    fn get_area(&self) -> f64 {
        // Simplified area calculation for the house
        let base_area = 100.0 * 60.0 * self.scale * self.scale;
        let roof_area = 0.5 * 120.0 * 50.0 * self.scale * self.scale;
        base_area + roof_area
    }

    fn get_info(&self) -> String {
        format!(
            "Complex Shape (House) at ({:.1}, {:.1}) with scale {:.2} using {}",
            self.center_x,
            self.center_y,
            self.scale,
            self.drawing_api.get_api_name()
        )
    }
}

/// Notification Bridge Pattern Example
pub trait NotificationSender {
    fn send(&self, message: &str, recipient: &str) -> Result<String, String>;
    fn get_sender_type(&self) -> &str;
}

pub struct EmailSender {
    smtp_server: String,
}

impl EmailSender {
    pub fn new(smtp_server: &str) -> Self {
        Self {
            smtp_server: smtp_server.to_string(),
        }
    }
}

impl NotificationSender for EmailSender {
    fn send(&self, message: &str, recipient: &str) -> Result<String, String> {
        if !recipient.contains('@') {
            return Err("Invalid email address".to_string());
        }
        Ok(format!(
            "Email sent via {} to {}: {}",
            self.smtp_server, recipient, message
        ))
    }

    fn get_sender_type(&self) -> &str {
        "Email"
    }
}

pub struct SmsSender {
    provider: String,
}

impl SmsSender {
    pub fn new(provider: &str) -> Self {
        Self {
            provider: provider.to_string(),
        }
    }
}

impl NotificationSender for SmsSender {
    fn send(&self, message: &str, recipient: &str) -> Result<String, String> {
        if recipient.len() < 10 {
            return Err("Invalid phone number".to_string());
        }
        Ok(format!(
            "SMS sent via {} to {}: {}",
            self.provider, recipient, message
        ))
    }

    fn get_sender_type(&self) -> &str {
        "SMS"
    }
}

pub struct SlackSender {
    workspace: String,
}

impl SlackSender {
    pub fn new(workspace: &str) -> Self {
        Self {
            workspace: workspace.to_string(),
        }
    }
}

impl NotificationSender for SlackSender {
    fn send(&self, message: &str, recipient: &str) -> Result<String, String> {
        if !recipient.starts_with('#') && !recipient.starts_with('@') {
            return Err("Invalid Slack recipient format".to_string());
        }
        Ok(format!(
            "Slack message sent in workspace {} to {}: {}",
            self.workspace, recipient, message
        ))
    }

    fn get_sender_type(&self) -> &str {
        "Slack"
    }
}

/// Notification abstraction
pub struct Notification {
    sender: Box<dyn NotificationSender>,
    priority: NotificationPriority,
}

#[derive(Debug, Clone)]
pub enum NotificationPriority {
    Low,
    Normal,
    High,
    Critical,
}

impl Notification {
    pub fn new(sender: Box<dyn NotificationSender>, priority: NotificationPriority) -> Self {
        Self { sender, priority }
    }

    pub fn send(&self, message: &str, recipient: &str) -> Result<String, String> {
        let priority_prefix = match self.priority {
            NotificationPriority::Low => "[LOW]",
            NotificationPriority::Normal => "[INFO]",
            NotificationPriority::High => "[HIGH]",
            NotificationPriority::Critical => "[CRITICAL]",
        };

        let formatted_message = format!("{} {}", priority_prefix, message);
        self.sender.send(&formatted_message, recipient)
    }

    pub fn get_info(&self) -> String {
        format!("{} notification with {:?} priority", self.sender.get_sender_type(), self.priority)
    }
}

/// Urgent notification - extends the abstraction
pub struct UrgentNotification {
    sender: Box<dyn NotificationSender>,
    max_retries: u32,
}

impl UrgentNotification {
    pub fn new(sender: Box<dyn NotificationSender>, max_retries: u32) -> Self {
        Self { sender, max_retries }
    }

    pub fn send_with_retry(&self, message: &str, recipient: &str) -> Result<String, String> {
        let urgent_message = format!("[URGENT] {}", message);

        for attempt in 1..=self.max_retries {
            match self.sender.send(&urgent_message, recipient) {
                Ok(result) => return Ok(format!("Attempt {}: {}", attempt, result)),
                Err(e) if attempt == self.max_retries => return Err(format!("Failed after {} attempts: {}", self.max_retries, e)),
                Err(_) => continue,
            }
        }

        Err("Unexpected error".to_string())
    }

    pub fn get_info(&self) -> String {
        format!("Urgent {} notification with {} max retries", self.sender.get_sender_type(), self.max_retries)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_circle_with_different_apis() {
        let canvas_circle = Circle::new(10.0, 20.0, 5.0, Box::new(CanvasAPI::new(800, 600)));
        let svg_circle = Circle::new(10.0, 20.0, 5.0, Box::new(SvgAPI::new(2)));

        let canvas_output = canvas_circle.draw();
        let svg_output = svg_circle.draw();

        assert!(canvas_output[0].contains("Canvas"));
        assert!(svg_output[0].contains("circle"));
        assert_eq!(canvas_circle.get_area(), svg_circle.get_area());
    }

    #[test]
    fn test_shape_operations() {
        let mut rect = Rectangle::new(0.0, 0.0, 10.0, 20.0, Box::new(CanvasAPI::new(800, 600)));

        let initial_area = rect.get_area();
        rect.resize(2.0);
        assert_eq!(rect.get_area(), initial_area * 4.0);

        rect.move_to(50.0, 60.0);
        assert!(rect.get_info().contains("50.0, 60.0"));
    }

    #[test]
    fn test_notification_bridge() {
        let email_notification = Notification::new(
            Box::new(EmailSender::new("smtp.gmail.com")),
            NotificationPriority::High
        );

        let result = email_notification.send("Test message", "user@example.com").unwrap();
        assert!(result.contains("[HIGH]"));
        assert!(result.contains("smtp.gmail.com"));

        let sms_notification = Notification::new(
            Box::new(SmsSender::new("Twilio")),
            NotificationPriority::Critical
        );

        let result = sms_notification.send("Alert", "1234567890").unwrap();
        assert!(result.contains("[CRITICAL]"));
        assert!(result.contains("Twilio"));
    }

    #[test]
    fn test_urgent_notification() {
        let urgent = UrgentNotification::new(
            Box::new(SlackSender::new("dev-team")),
            3
        );

        let result = urgent.send_with_retry("System down!", "@channel").unwrap();
        assert!(result.contains("Attempt"));
        assert!(result.contains("[URGENT]"));
    }

    #[test]
    fn test_drawing_api_differences() {
        let canvas_api = CanvasAPI::new(800, 600);
        let svg_api = SvgAPI::new(3);
        let opengl_api = OpenGLAPI::new("4.0");

        let canvas_circle = canvas_api.draw_circle(10.0, 10.0, 5.0);
        let svg_circle = svg_api.draw_circle(10.0, 10.0, 5.0);
        let opengl_circle = opengl_api.draw_circle(10.0, 10.0, 5.0);

        assert!(canvas_circle.contains("Canvas"));
        assert!(svg_circle.contains("<circle"));
        assert!(opengl_circle.contains("OpenGL"));
    }
}

/// Example usage and demonstration
pub fn demo() {
    println!("=== Bridge Pattern Demo ===");

    // Drawing API Bridge example
    println!("\n1. Drawing API Bridge:");

    let apis: Vec<Box<dyn DrawingAPI>> = vec![
        Box::new(CanvasAPI::new(800, 600)),
        Box::new(SvgAPI::new(2)),
        Box::new(OpenGLAPI::new("4.0")),
    ];

    for (i, api) in apis.into_iter().enumerate() {
        println!("\nUsing {}:", api.get_api_name());

        // Create shapes with different APIs
        let mut circle = Circle::new(100.0, 100.0, 50.0, api);
        println!("  {}", circle.get_info());

        let draw_commands = circle.draw();
        for command in draw_commands {
            println!("  Command: {}", command);
        }

        circle.resize(1.5);
        circle.move_to(150.0, 150.0);
        println!("  After resize and move: {}", circle.get_info());
        println!("  Area: {:.2}", circle.get_area());
    }

    // Complex shape example
    println!("\n2. Complex Shape Example:");
    let mut house = ComplexShape::new(200.0, 200.0, Box::new(SvgAPI::new(1)));

    println!("Drawing house:");
    let house_commands = house.draw_house();
    for (i, command) in house_commands.iter().enumerate() {
        println!("  {}: {}", i + 1, command);
    }

    house.resize(0.8);
    house.move_to(300.0, 250.0);
    println!("House after resize and move: {}", house.get_info());

    // Notification Bridge example
    println!("\n3. Notification System Bridge:");

    let senders: Vec<Box<dyn NotificationSender>> = vec![
        Box::new(EmailSender::new("smtp.company.com")),
        Box::new(SmsSender::new("Twilio")),
        Box::new(SlackSender::new("engineering")),
    ];

    let priorities = vec![
        NotificationPriority::Normal,
        NotificationPriority::High,
        NotificationPriority::Critical,
    ];

    for (sender, priority) in senders.into_iter().zip(priorities.into_iter()) {
        let notification = Notification::new(sender, priority);
        println!("\n{}", notification.get_info());

        let recipients = match notification.sender.get_sender_type() {
            "Email" => vec!["user@example.com", "admin@example.com"],
            "SMS" => vec!["1234567890", "0987654321"],
            "Slack" => vec!["@user", "#general"],
            _ => vec!["unknown"],
        };

        for recipient in recipients {
            match notification.send("Server maintenance scheduled", recipient) {
                Ok(result) => println!("  ✓ {}", result),
                Err(error) => println!("  ✗ {}", error),
            }
        }
    }

    // Urgent notification example
    println!("\n4. Urgent Notification Example:");
    let urgent_email = UrgentNotification::new(
        Box::new(EmailSender::new("emergency.smtp.com")),
        3
    );

    println!("{}", urgent_email.get_info());
    match urgent_email.send_with_retry("Critical system failure detected!", "ops@company.com") {
        Ok(result) => println!("✓ {}", result),
        Err(error) => println!("✗ {}", error),
    }

    // Cross-platform drawing example
    println!("\n5. Cross-Platform Drawing:");
    let shapes_with_apis: Vec<(Box<dyn Shape>, &str)> = vec![
        (
            Box::new(Rectangle::new(0.0, 0.0, 100.0, 50.0, Box::new(CanvasAPI::new(1024, 768)))),
            "Canvas Rectangle"
        ),
        (
            Box::new(Circle::new(50.0, 50.0, 25.0, Box::new(SvgAPI::new(3)))),
            "SVG Circle"
        ),
        (
            Box::new(Rectangle::new(0.0, 0.0, 80.0, 80.0, Box::new(OpenGLAPI::new("3.3")))),
            "OpenGL Rectangle"
        ),
    ];

    for (shape, description) in shapes_with_apis {
        println!("\n{}:", description);
        println!("  Info: {}", shape.get_info());
        println!("  Area: {:.2}", shape.get_area());

        let commands = shape.draw();
        for command in commands {
            println!("  Draw: {}", command);
        }
    }
}

fn main() {
    demo();
}